askrene: improving the MCF solver

Lagrang3 opened the pull request:
This PR aims to improve the MCF solver.

The idea is to have an API that solves specifically the MCF and related problem variations
at a lower level that do not have to care about the several details that are inherent in our modeling of the payment problem.

This approach allows me to easily change the algorithms in the backend.
Now we are using the same linearization of the channel cost function that it is used in `renepay`.
Future improvements could include the treatment of additional constraints and activation costs to arcs (https://github.com/Lagrang3/mcf-algorithms).

In the translation from a lightning network payment problem to a graph theoretical optimization problem 
I have added the possibility to choose the minimum unit of account, not necessarily msat but could be any other number
depending on the total amount to pay.

rustyrussell commented on plugins/askrene/graph.c:
```diff
@@ -0,0 +1,79 @@
+	graph->node_adjacency_next =
+	    tal_arr(graph, struct arc, graph->max_num_arcs);
+
+	/* bad allocation of graph components */
+	if (!graph->arc_tail || !graph->node_adjacency_first ||
+	    !graph->node_adjacency_next) {
+		return tal_free(graph);
+	}
```
tal never fails (well, it calls abort internally), and anyway this path is untested, so generally we simplify these checks out.

Lagrang3 replied:
I always check for allocation. If it is ok to ignore it I can remove all those checks.

Lagrang3 replied:
Fixed f00c63fe457f1d0e9f47bcc7732aebca7e919984

rustyrussell commented on plugins/askrene/graph.h:
```diff
@@ -0,0 +1,165 @@
+static inline struct node arc_tail(const struct graph *graph,
+				   const struct arc arc)
+{
+	assert(arc.idx < tal_count(graph->arc_tail));
```
I suspect (but have not measured!) that this will introduce a perf regression.  Usually we don't care, but these are called so often that I suggest:

1. use graph_max_num_arcs()
2. Implement graph_check() which does these assertions, and call it in various places (e.g. end of constructor)

Lagrang3 replied:
I remember now the small performance tweak by avoiding call to `tal_count`. I can fix it.
But shouldn't we measure performance with release type of builds anyways? ie. compiled with -O2/3 and NDEBUG.

Lagrang3 replied:
Changed `tal_count` for `graph_max_num_arcs/nodes` in checks. d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -0,0 +1,79 @@
+	for (size_t i = 0; i < max_num_nodes; i++)
+		prev[i].idx = INVALID_INDEX;
+
+	LQUEUE(struct queue_data, ql) myqueue = LQUEUE_INIT;
+	struct queue_data *qdata;
+
+	qdata = tal(this_ctx, struct queue_data);
+	qdata->idx = source.idx;
+	lqueue_enqueue(&myqueue, qdata);
```
This reverts an important optimization made in 9f0c0e1ccab9283799a47f021779fe45bb30b167 ?

Lagrang3 replied:
d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -77,3 +77,95 @@ bool BFS_path(const tal_t *ctx, const struct graph *graph,
+	bool target_found = false;
+	const size_t max_num_arcs = graph_max_num_arcs(graph);
+	const size_t max_num_nodes = graph_max_num_nodes(graph);
+	tal_t *this_ctx = tal(ctx, tal_t);
```
const...

Lagrang3 replied:
d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -77,3 +77,95 @@ bool BFS_path(const tal_t *ctx, const struct graph *graph,
+	tal_t *this_ctx = tal(ctx, tal_t);
+
+	/* check preconditions */
+	if (!graph || source.idx >=max_num_nodes || !cost || !capacity ||
+	    !prev || !distance)
+		goto finish;
+
```
Assert, don't fail gracefully here.  Make sure caller never does this.  Once it's in the API as "permitted" it's really hard to remove.

Lagrang3 replied:
d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -77,3 +77,95 @@ bool BFS_path(const tal_t *ctx, const struct graph *graph,
+	bitmap *visited = tal_arrz(this_ctx, bitmap,
+				   BITMAP_NWORDS(max_num_nodes));
+
+	if (!visited)
+		/* bad allocation */
+		goto finish;
```
allocations don't fail

Lagrang3 replied:
Fixed f00c63fe457f1d0e9f47bcc7732aebca7e919984

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -322,3 +322,87 @@ s64 node_balance(const struct graph *graph,
+
+	if (tal_count(capacity) != max_num_arcs ||
+	    tal_count(cost) != max_num_arcs)
+		goto finish;
```
Again, I prefer crashing.

Lagrang3 replied:
d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -212,11 +212,35 @@ static s64 get_augmenting_flow(const struct graph *graph,
+ * Sends an amount of flow through an arc, changing the flow balance of the
+ * nodes connected by the arc and the [residual] capacity of the arc and its
+ * dual. */
+static inline void sendflow(const struct graph *graph, const struct arc arc,
```
inline is generally not a good idea in C files.  In fact, it's almost certainly ignored.

Lagrang3 replied:
d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -341,68 +341,329 @@ s64 node_balance(const struct graph *graph,
+	struct node target = {.idx = INVALID_INDEX};
+	tal_t *this_ctx = tal(ctx, tal_t);
+
+	if (!this_ctx)
+		/* bad allocation */
+		goto finish;
```
Unnecessary

Lagrang3 replied:
Fixed f00c63fe457f1d0e9f47bcc7732aebca7e919984

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -341,68 +341,329 @@ s64 node_balance(const struct graph *graph,
+					 s64 *distance)
+{
+	struct node target = {.idx = INVALID_INDEX};
+	tal_t *this_ctx = tal(ctx, tal_t);
```
const?

Lagrang3 replied:
d24e60a643e377a91163afb0271d42958b04d027

rustyrussell commented on plugins/askrene/algorithm.c:
```diff
@@ -341,68 +341,329 @@ s64 node_balance(const struct graph *graph,
+		prev[i].idx = INVALID_INDEX;
+
+/* Only in debug mode we keep track of visited nodes. */
+#ifndef NDEBUG
+	bitmap *visited =
+	    tal_arrz(this_ctx, bitmap, BITMAP_NWORDS(max_num_nodes));
+	assert(visited);
+#endif
```
We don't generally use NDEBUG in our code (it turns off your assert()s above, too).  So this probably needs to be a runtime flag, like dev_askrene_debug, and maybe make it a global set if --developer is true?

Lagrang3 replied:
OK. Then I will remove this check which is unnecessary.

rustyrussell commented on plugins/askrene/test/run-mcf-large.c:
```diff
@@ -0,0 +1,111 @@
+int main(int argc, char *argv[])
+{
+	common_setup(argv[0]);
+	freopen("plugins/askrene/test/data/linear_mcf", "r", stdin);
```
Use zlib here?  That takes data down from 256k to 90k, which is  nice.  If we don't have zlib, just exit 0.

Lagrang3 replied:
Done 3f39522075cdb6dc1c4befdd643748b55ba7101f

rustyrussell commented on plugins/askrene/test/run-mcf-large.c:
```diff
@@ -22,6 +22,10 @@
 
+/* check-source does not like me to use scanf, how am I going to read the test
+ * data otherwise? */
+#define myscanf scanf
```
LoL.  fscanf :)


But with zlib you'll end up with `gzgets(gzFile file, char *buf, int len);` then use sscanf on the buffer anyway.

Lagrang3 replied:
3f39522075cdb6dc1c4befdd643748b55ba7101f

rustyrussell commented on common/amount.h:
```diff
@@ -148,6 +148,12 @@ bool amount_msat_eq_sat(struct amount_msat msat, struct amount_sat sat);
 double amount_msat_ratio(struct amount_msat a, struct amount_msat b);
 
+/* floor(a/b) */
+u64 amount_msat_ratio_floor(struct amount_msat a, struct amount_msat b);
```
u64?  I'd expect these to return double?

Lagrang3 replied:
Floor(x): the biggest integer that is not greater than x.
It is integer anyways.

rustyrussell reviewed (commented):
Minor tweaks.

Mainly concerned about performance regressions, and of course those test need re-enabling and fixing up (of course there will be differences, but the tests are there to ensure changes are *deliberate* at least!).

Would be interested in times of test_askrene.py::test_real_data before and after...

Lagrang3 commented:
Measuring time for every `minflow` call while running `test_real_data`, eg.
```
	t1 = time_now().ts;
	minflow(...)
	t2 = time_now().ts;
```
There are several calls to `minflow` in that test case. I measured the time for every `minflow` and put all those values into a histogram. I see no big difference in performance.
![Screenshot from 2024-11-13 21-22-44](https://github.com/user-attachments/assets/ecd4566c-4448-4879-816b-9e926a9f428f)

Lagrang3 commented:
Zooming in it looks there is a small gain in performance overall
![Screenshot from 2024-11-13 21-31-46](https://github.com/user-attachments/assets/f5c09d3e-0572-460b-9621-dd2d7ebdf8a6)

Lagrang3 commented:
Rebased on top of the master branch to fix the conflicts.
After playing with the probability cost factor a little bit I have come back to have `k = get_median_ratio()`,
because it gives better results in `test_real_data` than `k = 1000 * get_median_ratio()`.

Still have 3 more test cases that fail:
- `test_getroutes_fee_fallback`: where the problem is again the probability cost factor, here the ratio of the median is not enough,
- `test_askrene_fake_channeld`: which also fails to me in the master branch,
- `test_real_biases`: which shouldn't fail, I think I messed up something with the biases when I rebased, I am investigating this right now.

Lagrang3 commented:
I have found one assertion in `test_real_biases` to be too strong to be too strong 
```
@@ -1192,7 +1191,7 @@ def test_real_biases(node_factory, bitcoind):
             if route2 != route:
                 # It should have avoided biassed channel
                 amount_after = amount_through_chan(chan, route2['routes'])
-                assert amount_after < amount_before
+                # assert amount_after < amount_before
                 num_changed[bias] += 1
```

Finding an alternative set of routes does not imply that the biassed channel has less amount allocated.
The assertion fails only on node 83.
One possible explanation is that there is a very good channel `C` to get to node 83, even if biassed by one factor.
The flow algorithm determines several routes to get to node 83 using that channel `C`, but the solution to the Min. Cost Flow
is not guaranteed to be unique, so we might end up with two different set of routes before and after applying the bias
where the liquidity that goes through `C` is always the same or even slightly different. But there is no guarantee
that the biased `C` will have strictly less allocation. It could be even be higher due to the limited precision of the base unit
and the way we handle base fees approximately.

Lagrang3 commented:
Rebased

rustyrussell commented:
> The fee_fallback test would fail after fixing the computation of the median. Now we can restore it by making the probability cost factor 1000x higher than the ratio of the median. This shows how hard it is to combine fee and probability costs and why is the current approach so fragile. @rustyrussell

Actually, it shows how bad this generate graph is!

In particular, the reverse edges (which all have tiny fees but are irrelevant), warp the median so badly the result is bogus (I ended up having to modify the code to dump these out, to see what was happening). Fix is simple: disable reverse edges.

I also hardcoded an exception for the test_real_biases fail.  It's actually the same-cost path, it's just that the slight change in values means we pay 1msat more due to fee rounding!

No, median sucks because of this:

```
   83.50%     0.00%  cln-askrene  cln-askrene           [.] listpeerchannels_done                                                    ▒
+   83.48%     0.00%  cln-askrene  cln-askrene           [.] do_getroutes                                                             ▒
+   83.47%     0.00%  cln-askrene  cln-askrene           [.] get_routes                                                               ▒
-   80.71%     0.00%  cln-askrene  cln-askrene           [.] minflow                                                                  ▒
   - 80.71% minflow                                                                                                                   ▒
      - 29.96% combine_cost_function                                                                                                  ▒
         - 28.45% get_median_ratio                                                                                                    ▒
            + 15.84% __GI___qsort_r                                                                                                   ▒
            + 11.26% __GI___qsort_r                                                                                                   ▒
      - 27.96% mcf_refinement                                                                                                         ▒
         - 26.75% dijkstra_nearest_sink                                                                                               ▒
            - 3.28% priorityqueue_update                                                                                              ▒
               + 2.63% gheap_restore_heap_after_item_increase                                                                         ▒
              2.77% node_adjacency_next                                                                                               ▒
              2.20% arc_head                                                                                                          ▒
            + 2.08% priorityqueue_pop                                                                                                 ▒
              0.70% arc_tail                                                                                                          ▒
      - 8.03% init_linear_network                                                                                                     ▒
           2.02% gossmap_nth_chan                                                                                                     ▒
         - 1.72% linearize_channel                                                                                                    ▒
              0.54% get_constraints                                                                                                   ▒
         - 1.68% graph_add_arc                                                                                                        ▒
              1.37% graph_push_outbound_arc                                                                                           ▒
      + 6.56% get_flow_paths                                                                                                          ▒
      + 6.40% simple_feasibleflow                                                                                                     ▒
        1.54% init_residual_network                                                                                                   ▒
```

We spend an annoying 3/8 of our time getting the median values!  Maybe champagne problems, since the rest is so fast?  But it annoys me!

rustyrussell commented:
Trivial rebase and new commit...
